src/OpenColorIO/ops/Lut3D/Lut3DOpData.cpp

```diff
@@ -0,0 +1,1048 @@
+    Lut3DOpDataRcPtr newDomain = std::make_shared<Lut3DOpData>(GridSize);
+
+    // Regardless of what depth is used to build the domain, set the in & out to the
+    // actual depth so that scaling is done correctly.
+    newDomain->setInputBitDepth(lut->getInputBitDepth());
+    newDomain->setOutputBitDepth(lut->getInputBitDepth());
+
+    // Compose the LUT newDomain with our inverse LUT (using INV_EXACT style).
+    Lut3DOpDataRcPtr newLut(Lut3DOpData::Compose(newDomain, lut));
+
+    // The EXACT inversion style computes an inverse to the tetrahedral
+    // style of forward evalutation.
+    // TODO: Although this seems like the "correct" thing to do, it does
+    // not seem to help accuracy (and is slower).  To investigate ...
+    //newLut->setInterpolation(SYNCOLOR::LUT3D_ALGO_TETRAHEDRAL);
+
+    return newLut;
+}
+
+const unsigned long Lut3DOpData::maxSupportedLength = 129;
+
+namespace
+{
+void EvalTransform(const float * in,
+                   float * out,
+                   long numPixels,
+                   OpRcPtrVec & ops)
+{
+    std::vector<float> tmp(numPixels * 4);
+
+    // Render the LUT entries (domain) through the ops.
+
+    const float * values = in;
+    for (long idx = 0; idx<numPixels; ++idx)
+    {
+        tmp[4 * idx + 0] = values[0];
+        tmp[4 * idx + 1] = values[1];
+        tmp[4 * idx + 2] = values[2];
+        tmp[4 * idx + 3] = 1.0f;
+
+        values += 3;
+    }
+
+    // NB: We assume finalize will set the bit-depth at each op interface
+    // to 32f so there is never any quantization to integer.
+    // Furthermore, to avoid improper recursion we must never call
+    // the renderer with an integer input depth.
+    // TODO: Currently the OCIO finalize is hard-coded to 32f.
+    //       When that changes we need to update this to request 32f input
+    //       and output bit depths.
+    FinalizeOpVec(ops);
+
+    for (OpRcPtrVec::size_type i = 0, size = ops.size(); i<size; ++i)
+    {
+        ops[i]->apply(&tmp[0], numPixels);
+    }
+
+    float * result = out;
+    for (long idx = 0; idx<numPixels; ++idx)
+    {
+        result[0] = tmp[4 * idx + 0];
+        result[1] = tmp[4 * idx + 1];
+        result[2] = tmp[4 * idx + 2];
+
+        result += 3;
+    }
+}
+}
+
+// Functional composition is a concept from mathematics where two functions
+// are combined into a single function.  This idea may be applied to ops
+// where we generate a single op that has the same (or similar) effect as
+// applying the two ops separately.  The motivation is faster processing.
+//
+// When composing LUTs, the algorithm produces a result which takes the
+// domain of the first op into the range of the last op.  So the algorithm
+// needs to render values through the ops.  In some cases the domain of
+// the first op is sufficient, in other cases we need to create a new more
+// finely sampled domain to try and make the result less lossy.
+Lut3DOpDataRcPtr Lut3DOpData::Compose(const Lut3DOpDataRcPtr & A,
```

Comment:
Yeah, a global flag would be an interesting possibility.

Reply:
Contrary to what Patrick and Bernard wrote, SynColor generally does NOT compose two consecutive 3d-LUTs.  For example, if a config has two ColorSpaces that each include a 3d-LUT and you create a processor to convert between them, then those two 3d-LUTs are applied individually, they are NOT composed.  (Same if you have a single ColorSpace that uses several 3d-LUTs, they would get applied individually, for accuracy.)

We plan to do some work on the optimization that happens once a transform gets finalized in some future PRs.  As part of that, we plan to add an option to compose adjacent 3d-LUTs but that will never be the default, only an option for users that prize speed over quality.  (SynColor as it is used today in Flame, Maya, etc. never does this type of 3d-LUT optimization.)

The reason we needed to pull the Compose function into this PR is that the exact inversion algorithm that is used on the CPU cannot really be ported to the GPU.  (An exact inversion would be possible on GPUs but it would need a different algorithm and some R&D.)  Hence the way we implement inversion on the GPU is to take an identity 3d-LUT and use Compose to merge it with the exact inversion (on the CPU) to make a single forward 3d-LUT that approximates the exact inverse.  In other words, Compose is only used to generate the "fast" inversion style.

Reply:
Thanks for the explanation Doug.

Reply:
In Maya, no UI from the Color Management Preferences provide a way to change the color management optimization level, and the default implementation unfortunately composes 3D LUTs. Only the env. variable SYNCOLOR_OPTIMIZER_VALUE (set before starting Maya) overwrites the default optimization level.

@michdolan  Such a basic question did in fact highlight a problem in Maya that will be fixed in a following release.